lib/BC/SleighLifter.cpp

```diff
@@ -0,0 +1,1292 @@
+                                false);
+        }
+      }
+    }
+
+    if (opc == OpCode::CPUI_PIECE && outvar) {
+      assert(rhs.size + lhs.size == outvar->size);
+
+      // Treat them as integers
+      auto lifted_lhs = this->LiftInParam(
+               bldr, lhs, llvm::IntegerType::get(this->context, lhs.size * 8)),
+           lifted_rhs = this->LiftInParam(
+               bldr, rhs, llvm::IntegerType::get(this->context, rhs.size * 8));
+
+      if (lifted_lhs.has_value() && lifted_rhs.has_value()) {
+        // Widen the most significant operand and then left shift it to make room for the least significant operand.
+        auto *ms_operand = bldr.CreateZExt(
+            *lifted_lhs, llvm::IntegerType::get(this->context, outvar->size));
+        auto *shifted_ms_operand = bldr.CreateShl(
+            ms_operand, llvm::ConstantInt::get(
+                            llvm::Type::getInt8Ty(this->context), rhs.size));
+
+        // Now concatenate them with an OR.
+        auto *concat = bldr.CreateOr(shifted_ms_operand, *lifted_rhs);
+        return std::make_pair(this->LiftStoreIntoOutParam(bldr, concat, outvar),
+                              false);
+      }
+    }
+
+    if (opc == OpCode::CPUI_SUBPIECE && outvar) {
+      auto lifted_lhs = this->LiftInParam(
+          bldr, lhs, llvm::IntegerType::get(this->context, lhs.size * 8));
+
+      if (lifted_lhs.has_value()) {
+        auto new_size = lhs.size - rhs.offset;
+        auto *subpiece_lhs = bldr.CreateTrunc(
+            *lifted_lhs, llvm::IntegerType::get(this->context, new_size * 8));
+
+        if (new_size < outvar->size) {
+          subpiece_lhs = bldr.CreateZExt(
+              subpiece_lhs,
+              llvm::IntegerType::get(this->context, 8 * outvar->size));
+        } else if (new_size > outvar->size) {
+          subpiece_lhs = bldr.CreateTrunc(
+              subpiece_lhs,
+              llvm::IntegerType::get(this->context, 8 * outvar->size));
+        }
+
+        return std::make_pair(
+            this->LiftStoreIntoOutParam(bldr, subpiece_lhs, outvar), false);
+      }
+    }
+
+    if (opc == OpCode::CPUI_INDIRECT && outvar) {
+      // TODO(alex): This isn't clear to me from the documentation.
+      // I'll probably need to find some code that generates this op in order to understand how to handle it.
+      return std::make_pair(LiftStatus::kLiftedUnsupportedInstruction, false);
+    }
+
+    if (opc == OpCode::CPUI_NEW && outvar) {
+      // NOTE(alex): We shouldn't encounter this op as it only get generated when lifting Java or
+      // Dalvik bytecode
+      return std::make_pair(LiftStatus::kLiftedUnsupportedInstruction, false);
+    }
+
+    return std::make_pair(LiftStatus::kLiftedUnsupportedInstruction, false);
+  }
+
+  LiftStatus LiftTernOp(llvm::IRBuilder<> &bldr, OpCode opc,
+                        VarnodeData *outvar, VarnodeData param0,
+                        VarnodeData param1, VarnodeData param2) {
+    switch (opc) {
```

Comment:
`param0-2` is not very helpful naming - it is a ternary op, those operands have meaning - `cond, true_val, false_val` for example.

Reply:
I think this is a bad function name. It is just handling operands that have three operands, it's not a ternary conditional

Reply:
Then take an `std::tuple` or `std::array`? If they have no special meaning by themselves, why do they even need names?